lib/spack/docs/packaging_guide.rst

```diff
@@ -2775,6 +2775,254 @@ Suppose a user invokes ``spack install`` like this:
+=========== =========================================== =====================
+``root``    The installation directory for the package  ``spec.prefix``
+``command`` An executable command for the package       | ``spec.name`` in
+                                                        | ``spec.prefix.bin``
+``headers`` A list of headers provided by the package   | All headers
+                                                          searched
+                                                        | recursively in
+                                                          ``spec.prefix``
+``libs``    A list of libraries provided by the package | ``lib{spec.name}``
+                                                          searched
+                                                        | recursively in
+                                                          ``lib``, ``lib64``,
+                                                        | then the rest of
+                                                          of ``spec.prefix``
+=========== =========================================== =====================
+
+Each of these can be customized by implementing the relevant attribute
+as a ``@property`` in the package's class:
+
+.. code-block:: python
+   :linenos:
+
+   class Foo(Package):
+       ...
+       @property
+       def libs(self):
+           # The library provided by Foo is libMyFoo.so
+           return find_libraries('libMyFoo', root=self.spec.prefix, recursive=True)
+
+A package may also provide a custom implementation of each attribute
+for the virtual packages it provides by implementing the
+``virtualpackagename_attributename`` property in the package's class.
+The implementation used is the first one found from:
+
+#. Specialized virtual: ``Package.virtualpackagename_attributename``
+#. Generic package: ``Package.attributename``
+#. Default
+
+The use of customized attributes is demonstrated in the next example.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Example: Customized attributes for virtual packages
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Consider a package ``foo`` that can optionally provide two virtual
+packages ``bar`` and ``baz``.  When both are enabled the installation tree
+appears as follows:
+
+.. code-block:: console
+
+   include/foo.h
+   include/bar/bar.h
+   lib64/libFoo.so
+   lib64/libFooBar.so
+   baz/include/baz/baz.h
+   baz/lib/libFooBaz.so
+
+The install tree shows that ``foo`` is providing the header ``include/foo.h``
+and library ``lib64/libFoo.so`` in it's install prefix.  The virtual
+package ``bar`` is providing ``include/bar/bar.h`` and library
+``lib64/libFooBar.so``, also in ``foo``'s install prefix.  The ``baz``
+package, however, is provided in the ``baz`` subdirectory of ``foo``'s
+prefix with the ``include/baz/baz.h`` header and ``lib/libFooBaz.so``
+library.  Such a package could implement the optional attributes as
```

Comment:
Can we make the example be based on a real use case (which I am sure exists)? By just reading the docs I don't understand which use case this PR covers (given that `mpi_libs` etc. already exist as properties).

Reply:
There's two use cases in particular but they can't really be implemented until the feature is available so having the docs show an actual package implementing it is a chicken/egg problem.  The two use cases that motivated the pr (although there can certainly be others) are:

1. A package providing multiple components in sub-directories of it's install prefix making each component available as a virtual package.
2. Using dummy packages for complex scenarios that virtuals don't allow for (think virtuals providing virtuals).

1 is the easiest to understand and why I used it for the example.  It would be particularly useful for simplifying the intel and oneapi packages but is also useful for upcoming changes to paraview, vtk, and vtkm packages (vtk provides vtkm, and paraview provides vtk and vtkm).

2 is the more immediate use case and is being used to solve a structural issue with mesa, glx, and osmesa.  Mesa provides both glx and osmesa in the same build, and both of those have valid implementations of opengl.  Currently though there's no way to have mesa respond appropriately for `spec['gl'].libs` since there's no way of knowing *which* of the two `gl` implementation is being queried.  So I've created dummy `osmesa` and `glx` packages to allow the following:
```python
class Mesa(MesonPackage):
   provides('libosmesa', when='+osmesa')
   provides('libglx', when='+glx')


class Osmesa(BundlePackage):
    depends_on('libosmesa')
    provides('gl')

    @property
    def home(self):
        return self.spec['libosmesa'].prefix

    @property
    def headers(self):
        return self.spec['libosmesa'].headers

    @property
    def libs(self):
        return find_libraries('libOSMesa', root=self.home, recursive=True)

    @property
    def gl_headers(self):
        return find_headers('GL/gl.h', root=self.home.include, recursive=False)


class Glx(BundlePackage):
    depends_on('libglx')
    provides('gl')

    @property
    def home(self):
        return self.spec['libglx'].prefix

    @property
    def headers(self):
        return self.spec['libglx'].headers

    @property
    def libs(self):
        return find_libraries('libGL', root=self.home, recursive=True)

    @property
    def gl_headers(self):
        return find_headers('GL/gl.h', root=self.home.include, recursive=False)
```

This effectively lets `mesa` provide two different `gl` packges in the same build

Reply:
Regarding the second case, `spec['gl'].headers` and `.libs` work as is but `.prefix` will give the install path of the dummy pacakge which is not particularly helpful.  `.home` will give the correct install prefix delegated to `mesa` even though `gl` is provided by `osmesa` or `glx`.